### PR TITLE
[exporter/clickhouse] Default `async_insert` to true. Added related config option.

### PR DESCRIPTION
**Pull request by SpencerTorres**
**Description:**
Sets `async_insert` to true by default to enable [asynchronous inserts](https://clickhouse.com/docs/en/optimize/asynchronous-inserts).
Because this value is being given a default, I have added a config option under the same name.
Keep in mind that if `async_insert` is provided in `endpoint` or `connection_params` it will take precedence and ignore this new config option.

This is similar to how the `database` config option behaves.
The goal is to provide better insert performance by default, since not all users will know to set it in their DSN URL.

This also opens the discussion to ___**whether or not this is a breaking change**___. Depending on the deployment's telemetry throughput, this could be an unexpected change that leads to [`TOO_MANY_PARTS`](https://clickhouse.com/docs/knowledgebase/exception-too-many-parts) errors. I don't expect this to be the case however, but I welcome any discussion about this concern.

**Link to tracking Issue:** <Issue number if applicable>

**Testing:**
Ran integration tests. Also added an abundance of tests to check the behavior of `async_insert` when present in `endpoint`, `connection_params`, and exporter config.

**Documentation:**
Added new config option to README + code comments.

Unrelated change, also updated README's SQL samples to use `sql` instead of `clickhouse` for the code samples to enable proper syntax highlighting. ClickHouse SQL is compatible with plain SQL.

### CODE REVIEW

**Review by Frapschen — APPROVED** [review]
LGTM

File: exporter/clickhouseexporter/config_test.go [hunk 1]
```diff
@@ -243,7 +244,76 @@ func TestConfig_buildDSN(t *testing.T) {
...
+			name: "use async_insert false config option when it is not present in DSN",
+			fields: fields{
+				Endpoint:    "tcp://127.0.0.1:9000",
+				AsyncInsert: configBool(false),
+			},
+
+			want: "tcp://127.0.0.1:9000/default?async_insert=false",
+		},
+		{
+			name: "set async_insert to true when not present in config or DSN",
```
Comment (crobert-1) [inline exporter/clickhouseexporter/config_test.go:311]: It would have to be done in its own test, but it would be good to add another test that asserts that the final endpoint is the same whether `async_insert` is coming through from the `endpoint` config option, from the `connection_params`, or from a combination of `async_insert` and `endpoint`. This way we can ensure going forward that the final result is the same from each configuration path.

File: exporter/clickhouseexporter/config.go [hunk 2]
```diff
@@ -47,6 +47,10 @@ type Config struct {
 	TableEngine TableEngine `mapstructure:"table_engine"`
 	// ClusterName if set will append `ON CLUSTER` with the provided name when creating tables.
 	ClusterName string `mapstructure:"cluster_name"`
+	// AsyncInsert if true will enable async inserts. Default is `true`.
+	// Ignored if async inserts are configured in the `endpoint` or `connection_params`.
+	// Async inserts may still be overridden server-side.
+	AsyncInsert *bool `mapstructure:"async_insert"`
```
[the remark this answers is not in the document]
Comment (SpencerTorres) [reply exporter/clickhouseexporter/config.go:53]: The default only gets applied if it's not in `endpoint`/`connection_params` AND if `cfg.AsyncInsert` is `nil`.

If it were a regular `bool`, Go would initialize the struct with `false`, and so in the config parsing, I wouldn't be able to determine if it was set to `false` by the user or by the struct's initialization.

So the code would look like:
```go
// This effectively ignores my config if I set cfg.AsyncInsert to false.
if cfg.AsyncInsert == false {
    queryParams.Set("async_insert", "true")
}
```

So if I wanted it to remain `false` in my config, it would never work, I would have to set it in `endpoint` or `connection_params`.

This is similar to the `database` config option, except in that case Go will initialize it to `""`, so the code checks for that instead of `nil`.


TL;DR: false is `false`, true is `true`, `nil` is `true`

File: exporter/clickhouseexporter/README.md [hunk 3]
```diff
@@ -283,6 +283,7 @@ Connection options:
 - `ttl` (default = 0): The data time-to-live example 30m, 48h. Also, 0 means no ttl.
 - `database` (default = otel): The database name.
 - `connection_params` (default = {}). Params is the extra connection parameters with map format.
+- `async_insert` (default = true): Enables [async inserts](https://clickhouse.com/docs/en/optimize/asynchronous-inserts). Ignored if async inserts are configured in the `endpoint` or `connection_params`. Async inserts may still be overridden server-side.
```
[the remark this answers is not in the document]
Comment (SpencerTorres) [reply exporter/clickhouseexporter/README.md:286]: I have added a note to the readme to clarify this.

`connection_params` will overwrite params in `endpoint` individually.

If `a` is in `endpoint` but `b` is in `connection_params`, the final URL will have `?a=1&b=1`. If `connection_params` also had `a` then the final URL would be `?a=1`.

In the context of this PR, `async_insert` would come after this process, so it only gets set if it's not present in either `endpoint` or `connection_params`

**Comment by SpencerTorres** [comment]
@crobert-1 I appreciate the review! I have applied all of your suggestions:
- Updated README note for `connection_params` overwriting `endpoint`
- Added test for `connection_params` overwriting `endpoint` and `async_insert`
- Updated changelog to your suggested phrase, as well as a sentence that says how to preserve the old behavior

Let me know if anything else stands out. Thanks!

File: exporter/clickhouseexporter/config.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (crobert-1) [reply exporter/clickhouseexporter/config.go:53]: Okay, I think I understand your concern. However, the way configurations are parsed by the collector works in this case. It first calls [createDefaultConfig](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/e870c7f0c05ec03e860df1a6a03e84dc0e308bb4/exporter/clickhouseexporter/factory.go#L32), then parses whatever configuration the user defined, overwriting any default values with given values (and leaving default values as-is if not defined by user). This is done [here](https://github.com/open-telemetry/opentelemetry-collector/blob/e8cabb785bb7bffd6c2e49f09563f132719495ea/otelcol/internal/configunmarshaler/configs.go#L41-L48), for reference.

So even if you have a `bool` with default `true`, if you set it to `true` in `createDefaultConfig`, it will only be overwritten if the user specifies `async_insert: false` in their config.

So would it work to use `bool` for your use case, or is there more reason than that?

[the remark this answers is not in the document]
Comment (crobert-1) [reply exporter/clickhouseexporter/config.go:53]: Hey @SpencerTorres, I was wondering if you got a chance to test this, or look into it more?
